ddl/db_partition_test.go

```diff
@@ -2087,6 +2087,190 @@ func TestGlobalIndexUpdateInDropPartition(t *testing.T) {
 	tk.MustQuery("select * from test_global use index(idx_b) order by a").Check(testkit.Rows("2 11 11", "12 12 12"))
 }
 
+func TestTruncatePartitionWithGlobalIndex(t *testing.T) {
+	defer config.RestoreFunc()()
+	config.UpdateGlobal(func(conf *config.Config) {
+		conf.EnableGlobalIndex = true
+	})
+
+	store := testkit.CreateMockStore(t)
+	tk := testkit.NewTestKit(t, store)
+	tk.MustExec("use test")
+	tk.MustExec("drop table if exists test_global")
+	tk.MustExec(`create table test_global ( a int, b int, c int)
+	partition by range( a ) (
+		partition p1 values less than (10),
+		partition p2 values less than (20)
+	);`)
+	tt := external.GetTableByName(t, tk, "test", "test_global")
+	pid := tt.Meta().Partition.Definitions[1].ID
+
+	tk.MustExec("Alter Table test_global Add Unique Index idx_b (b);")
+	tk.MustExec("Alter Table test_global Add Unique Index idx_c (c);")
+	tk.MustExec(`INSERT INTO test_global VALUES (1, 1, 1), (2, 2, 2), (11, 3, 3), (12, 4, 4)`)
+
+	tk.MustExec("alter table test_global truncate partition p2;")
+	result := tk.MustQuery("select * from test_global;")
+	result.Sort().Check(testkit.Rows(`1 1 1`, `2 2 2`))
```

Comment:
If one adds to this test, then in a new transaction in 'delete only' state, one cannot find the truncated "15 15 15" but it gives a duplicate error on insert.
```suggestion
       tk.MustExec(`INSERT INTO test_global VALUES (1, 1, 1), (2, 2, 2), (11, 3, 3), (12, 4, 4), (15, 15, 15)`)
 
       tk2 := testkit.NewTestKit(t, store)
       tk2.MustExec(`use test`)
       tk2.MustExec(`begin`)
       tk2.MustExec(`insert into test_global values (5,5,5)`)
       syncChan := make(chan bool)
       go func() {
               tk.MustExec("alter table test_global truncate partition p2;")
               syncChan <- true
       }()
       waitFor := func(i int, s string) {
               for {
                       tk4 := testkit.NewTestKit(t, store)
                       tk4.MustExec(`use test`)
                       res := tk4.MustQuery(`admin show ddl jobs where db_name = 'test' and table_name = 'test_global' and job_type = 'truncate partition'`).Rows()
                       if len(res) == 1 && res[0][i] == s {
                               break
                       }
                       time.Sleep(10 * time.Millisecond)
               }
       }
       waitFor(4, "delete only")
       tk3 := testkit.NewTestKit(t, store)
       tk3.MustExec(`begin`)
       tk3.MustExec(`use test`)
       tk3.MustQuery(`select b from test_global use index(idx_b) where b = 15`).Check(testkit.Rows())
       tk3.MustQuery(`select c from test_global use index(idx_c) where c = 15`).Check(testkit.Rows())
       // Here it will fail with
       // [kv:1062]Duplicate entry '15' for key 'test_global.idx_b'
       tk3.MustExec(`insert into test_global values (15,15,15)`)
       tk2.MustExec(`commit`)
       tk3.MustExec(`commit`)
       syncChan <- true
       result := tk.MustQuery("select * from test_global;")
       result.Sort().Check(testkit.Rows(`1 1 1`, `2 2 2`, `15 15 15`))
```

It can be quite confusing for a user if the row cannot be found in a transaction, while it still is not possible to insert.

Reply:
You are right. Because the global index row still exists when the state is in deleteOnly.

Maybe I should add write check for the truncating definitions, otherwise it will cause data inconsistency(imagine the global index records we just add will be deleted later). 

@mjonss What's your advice?

Reply:
@mjonss  I add write check when truncate partition ddl state is not in public.